rs/boundary_node/certificate_issuance/certificate_orchestrator/src/work.rs

```diff
@@ -178,6 +179,71 @@ impl<T: Peek> Peek for WithMetrics<T> {
     }
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum ListError {
+    #[error("Unauthorized")]
+    Unauthorized,
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+pub trait List {
+    fn list(&self) -> Result<Vec<(String, u64, Registration)>, ListError>;
+}
+
+pub struct Lister {
+    tasks: LocalRef<PriorityQueue<Id, Reverse<u64>>>,
+    registrations: LocalRef<StableMap<StorableId, Registration>>,
+}
+
+impl Lister {
+    pub fn new(
+        tasks: LocalRef<PriorityQueue<Id, Reverse<u64>>>,
+        registrations: LocalRef<StableMap<StorableId, Registration>>,
+    ) -> Self {
+        Self {
+            tasks,
+            registrations,
+        }
+    }
+}
+
+impl List for Lister {
+    fn list(&self) -> Result<Vec<(String, u64, Registration)>, ListError> {
+        self.tasks.with(|tasks| {
+            tasks
+                .borrow()
+                .iter()
+                .map(|(id, Reverse(timestamp))| {
+                    match self
+                        .registrations
+                        .with(|rs| rs.borrow().get(&id.to_owned().into()))
+                    {
+                        Some(r) => Ok((id.to_owned(), timestamp.to_owned(), r)),
+                        None => {
+                            Err(anyhow!("invalid state: task id not found in registrations").into())
```

Comment:
My thinking here was that if a registration wasn't found, that means that the canister is in some corrupt state (because a task should always have a corresponding registration) and so we should return an error instead of failing gracefully.

Reply:
My suggestions was just to include the ID of the missing registration in the error message. Something like this:
```suggestion
                            Err(anyhow!(format!("invalid state: task id {:?} not found in registrations", id)).into())
```